python/hidet/ir/Grammar.md

```diff
@@ -0,0 +1,101 @@
+### Notes
+Terminals: `t`
+Non-terminals: t
+
+t?: one or more
+t*: zero or more
+
+ALL_CAPS: Predefined terminal
+- STRING: A string enclosed by `""`
+- IDENT: A valid identifier in C/Python
+- INT/FLOAT/STRING: A valid literal in Python
+
+$\epsilon$ : Null
+( t | $\epsilon$ ) : Optional
+
+#### Minutiae
+- `@` to prevent ambiguity with Call
+- Every statement has an optional `#` IDENT, which encodes an optional attribute dict, the same attribute dict is defined in the top level module space
+- We encode things that are not part of the grammar as an attribute, for example, layouts. This keeps everything relatively extensible
+
+## Grammar
+Value := STRING | INT | FLOAT | Dict | List | Tuple
+Pair := STRING `:` Value
+Tuple := `(` ( Value | $\epsilon$ ) (`,` Value)\* `)`
+List := `[` ( Value | $\epsilon$ ) (`,` Value)\* `]`
+Dict := `{` ( Pair | $\epsilon$ ) (`,` Pair)\* `}`
+
+AttributeName := `#` IDENT
+Attribute := AttributeName `=` Dict `;`
+
+DataType := `i1` | `i8` | `i16` | `i32` | `i64` | `u8` | `u16` | `u32` | `u64` | `f16` | `f32` | `f64` | `void` | `str`
+PtrType := `~`DataType
+TensorType := DataType`<` NAT (`,` NAT)\* (`,` AttributeName | $\epsilon$ ) `>`
+TensorPtrType := `~`TensorType
+Type := DataType | PtrType | TensorType | TensorPtrType
+
+Module := `{` ( Function | Attribute )\* `}` 
+Function := `def` IDENT `(` ( Arg`:` Type | $\epsilon$ ) (`,` Arg`:` Type)\*` )` `->` Type `{` Stmt* `}`
```

Comment:
```suggestion
Function := `def` IDENT `(` ( Arg`:` Type | $\epsilon$ ) (`,` Arg`:` Type)* `)` `->` Type `{` Stmt* `}`
```

Reply:
I think that the escape sequence wasn't working before, I fixed it.